### PR TITLE
Fix panic on renewing a renewable KV v1 secret

### PR DESCRIPTION
**Pull request by maxb**
A regression was introduced in #17, in 2018. Better late than never,
let's fix it. This will open the door to considering deleting the second
copy of the KV v1 source code within the Vault repo, and just referring
to this as the canonical copy.

The fix selectively reverts some parts of #17, renames `handleRead` to
`handleReadOrRenew`, and adds some detailed comments to explain the
subtlety.

### CODE REVIEW

File: passthrough.go [hunk 1]
```diff
@@ -28,18 +28,18 @@ type Passthrough interface {
...
 func LeasedPassthroughBackendFactory(ctx context.Context, conf *logical.BackendConfig) (logical.Backend, error) {
-	return LeaseSwitchedPassthroughBackend(ctx, conf, true)
+	return LeaseSwitchedPassthroughBackendFactory(ctx, conf, true)
 }
 
-// LeaseSwitchedPassthroughBackend returns a PassthroughBackend
+// LeaseSwitchedPassthroughBackendFactory returns a PassthroughBackend
 // with leases switched on or off
-func LeaseSwitchedPassthroughBackend(ctx context.Context, conf *logical.BackendConfig, leases bool) (logical.Backend, error) {
+func LeaseSwitchedPassthroughBackendFactory(ctx context.Context, conf *logical.BackendConfig, leases bool) (logical.Backend, error) {
```
Comment (maxb) [inline passthrough.go:42]: Note to reviewer: The changes above are just an opportunistic cleanup whilst I was here ... one of the three factory functions was inconsistently missing the `Factory` suffix.

Compatibility: Should not really be an issue, no-one other than Vault should be importing this as a library and Vault doesn't (yet) call these.

File: passthrough_test.go [hunk 2]
```diff
@@ -270,6 +270,40 @@ func TestPassthroughBackend_Revoke(t *testing.T) {
...
+		t.Fatalf("err: %v", err)
+	}
+
+	req = logical.TestRequest(t, logical.RenewOperation, "foo")
+	req.Storage = storage
+	req.Secret = &logical.Secret{
+		InternalData: map[string]interface{}{
+			"secret_type": "kv",
+		},
+	}
```
Comment (maxb) [inline passthrough_test.go:292]: Note to reviewer: This is a bit fake, hacking together just enough of what a real RenewOperation request would look like. Ideally I'd like to run a real Core and generate a real lease renewal, but the relevant test helpers are not exposed for out-of-tree plugins to use. This is probably good enough, I guess.

**Review by maxb — COMMENTED** [review]
Note to reviewer: I made a conscious decision to not update the method name to `handleReadOrRenew` in this interface: https://github.com/hashicorp/vault-plugin-secrets-kv/blob/4eb796ea95e4828806dc457d18ed38800ff9d0bf/passthrough.go#L20-L26
because I already have #117 open to delete the interface entirely as unused code.

**Comment by maxb** [comment]
Hi @hsimon-hashicorp ... it's me again. I wonder if you'd be able to help get the right eyes on this one too? I fear it's a bit out of the way, in a rather quiet plugin repository - but it would be great to get it merged to unlock additional cleanups in the main Vault repository.

**Comment by raymonstah** [comment]
@maxb Hey Max, thanks for the contribution. Would you mind resolving the current merge conflicts? I'll take a look at the PR, but I'm not super familiar and don't have all of the context. I can try to help get more eyes on it.

**Comment by hsimon-hashicorp** [comment]
> @maxb Hey Max, thanks for the contribution. Would you mind resolving the current merge conflicts? I'll take a look at the PR, but I'm not super familiar and don't have all of the context. I can try to help get more eyes on it.

Thank you @raymonstah! I missed the original ping, and I'll see if we can get some eyes on it too.

**Comment by maxb** [comment]
> I'll take a look at the PR, but I'm not super familiar and don't have all of the context.

Thanks! In essence, the context is this:

* There is a rare operational mode for KVv1s where they are configured with leased_passthrough=true **and** individual secrets have a `ttl` key
* This was inadvertently broken in #17, causing Vault to panic during request handling if it is used
* Since it has been broken for some time, we may infer not many users are affected - although there is another report of the issue in #55 - however the broken functionality is **extensively** relied upon in Vault's own tests ... where it works because the Vault repo embeds an old **copy** of the KVv1 implementation, which is a bit inelegant, and sometimes a source of confusion
* The fix is pretty simple... selectively revert the unnecessary changes that broke it
* But I'm also adding comments, renaming a function, and adding a test, for insurance against this behaviour being overlooked again in the future

**Review by robmonte — APPROVED** [review]
Thanks for fixing and contributing this, I apologize for delays experienced with getting this merged.

**Pull request merged**